clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -423,18 +422,28 @@ else if (strategy == OffsetResetStrategy.LATEST)
 
                 nextInLineRecords = parseFetchedData(completedFetch);
             } else {
-                recordsRemaining -= append(drained, nextInLineRecords, recordsRemaining);
+                // check if we've already fetched data for this partition. If so, just return immediately
+                // to avoid the need to recopy the records to a new collection. This case should be rare
+                // because we do not send more than one fetch for a partition at a time.
+
+                TopicPartition partition = nextInLineRecords.partition;
+                if (drained.containsKey(partition))
+                    break;
```

Comment:
I could be persuaded either way. Really this case should never happen, but I said "rare" since it's difficult to rule out completely because of leader changes and such. In short, I don't think it matters too much how we handle it, so I opted for the simpler one.